### ISSUE TITLE
Understand and apply CapsNet on Traffic sign classification

### ISSUE THREAD

**Issue by sweaterr** [sweaterr/readinglist#327]
https://becominghuman.ai/understand-and-apply-capsnet-on-traffic-sign-classification-a592e2d4a4ea

**Comment by sweaterr** [comment]
I was not really convinced by relaunching a new test of neural network on the inevitable MNIST.
필연적 인 MNIST에 신경 네트워크의 새로운 테스트를 다시 시작함으로써 나는 정말로 확신하지 못했습니다.

But everyone seemed pretty excited about this new neural network architecture: CapsNet.
그러나 모두는이 새로운 신경 네트워크 아키텍처에 대해 매우 흥분해 보였습니다 : CapsNet.

Then I got interested too.
그럼 나도 관심있어.

As I struggle to just read a blog post when it comes to AI, I could not help but put my hands in the grease.
인공 지능에 관해서는 블로그 게시물을 읽는 데 어려움을 겪었으므로 도움을 청할 수는 없었습니다.

In addition, I was precisely working a self-driving car project (Udacity self driving car nanodegree) on a traffic sign dataset.
또한 교통 표지 데이터 세트에서자가 운전 자동차 프로젝트 (Udacity 자체 운전 차량 나노도)를 정확하게 작업했습니다.

That’s good !.
좋습니다!

**Comment by sweaterr** [comment]
Some useful links before you start:  The project’s github is located here.
시작하기 전에 유용한 링크 : 프로젝트의 github이 여기에 있습니다.

The project is based on Sara Sabour, Nicholas Frosst, and Geoffrey E.
이 프로젝트는 Sara Sabour, Nicholas Frosst 및 Geoffrey E.에 기반을두고 있습니다.

Hinton’s “Dynamic Routing Between Capsules” research paper.
Hinton의 "Dynamic Routing Between Capsules"연구 논문.

**Comment by sweaterr** [comment]
Before starting to criticize convolutional networks, I would like to point out that today, they are the state of the art in terms of image analysis and classification.
길쌈 네트워크를 비판하기 시작하기 전에, 나는 오늘날 그들이 이미지 분석과 분류 측면에서 최첨단임을 지적하고자한다.

CapsNet is a brilliant idea, certainly, but we will have to wait for Benchmarks more consistent than MNIST to really make an idea.
CapsNet은 훌륭한 아이디어이지만, 우리는 MNIST가 벤치마킹보다 더 일관성있게 아이디어를 내릴 때까지 기다려야 할 것입니다.

I am thinking of Imagenet among others.
저는 다른 사람들 중에서도 Imagenet을 생각하고 있습니다.

Nothing is played for the moment, even if beautiful things are to predict.
아름다운 일이 예언 되더라도 순간에는 아무 것도 연주되지 않습니다.

**Comment by sweaterr** [comment]
### What’s wrong with convolutional neural network?

Current CNN issues are partly based on their ability to generalize perceived images.
현재의 CNN 문제는 부분적으로 인식 된 이미지를 일반화하는 능력에 기반합니다.

For example : The rotation.
예 : 회전.

A network trained to recognize a type of image could mistake to recognize the same type of image with a certain rotation, which is why the training is often accompanied by data augmentation and Average / Max Pooling.
일종의 이미지를 인식하도록 훈련 된 네트워크는 실수로 특정 회전이있는 동일한 유형의 이미지를 인식 할 수 있으므로 교육에 종종 데이터 증가 및 평균 / 최대 풀링이 수반됩니다.

Pooling is a method of popularizing a layer by arbitrarily selecting a neurons subset for analysis on the lower layer.
풀링은 임의로 하위 계층 분석을위한 뉴런 하위 집합을 선택하여 계층을 대중화하는 방법입니다.

This has the advantage of making the next layers less expensive in computing power, but also to make the network less dependent on the original positions of the features in the image.
이는 컴퓨팅 성능면에서 다음 계층의 비용을 낮추는 장점이 있지만 이미지의 원래 위치에서 네트워크를 덜 종속적으로 만들 수 있다는 장점이 있습니다.

It is therefore assumed that the exact positions of the characteristics are not necessarily important to consider whether an object is present in the image or not.
따라서 특성의 정확한 위치가 물체가 이미지에 존재하는지 여부를 고려하는 것이 반드시 중요하지 않다고 가정합니다.

**Comment by sweaterr** [comment]
This allows the model to be invariant to the small change in images.
이렇게하면 모델이 이미지의 작은 변화에 대해 변하지 않게됩니다.

On the other hand, it is not really able to understand that some translation has taken place.
다른 한편으로, 어떤 번역이 일어났다는 것을 실제로 이해할 수는 없습니다.

Invariant means that no matter if the order and position of the detected characters change, the output will always be the same.
불변 식은 감지 된 문자의 순서와 위치가 변경 되더라도 출력은 항상 동일 함을 의미합니다.

But the perfect result would be rather to think in terms of equivariant.
그러나 완벽한 결과는 오히려 동등한 것으로 생각하는 것입니다.

The model would thus be able to understand rotations and positions of the features in the image, in order to produce an adequate output.
따라서 모델은 적절한 출력을 생성하기 위해 이미지에서 피처의 회전 및 위치를 이해할 수 있습니다.

Which is not really possible with the use of Pooling.
풀링을 사용하면 실제로 불가능합니다.

That’s what leads us to this new architecture.
그것이 우리를이 새로운 아키텍처로 안내합니다.

**Comment by sweaterr** [comment]
## CapsNet

A Capsulet network gives the model the ability to understand the changes in the image, and thus to better generalize what is perceived.
캡슐 렛 (Capsulet) 네트워크는 모델이 이미지의 변화를 이해하고 인식 된 것을 더 잘 일반화 할 수있는 능력을 부여합니다.

To understand how this architecture works, the important thing is to grasp the concept of Capsule.
이 아키텍처가 어떻게 작동하는지 이해하려면 중요한 점은 캡슐의 개념을 이해하는 것입니다.

“A capsule is a group of neurons whose activity vector represents the instantiation parameters of a specific type of entity such as an object or object part”.
"캡슐은 활동 벡터가 대상 또는 대상 부분과 같은 특정 유형의 개체의 인스턴스화 매개 변수를 나타내는 뉴런 그룹입니다."

We are used to talk about Deep Learning in terms of depth, this method introduces the notion of nesting which induces a new dimension of depth.
우리는 깊이에 관해 깊은 학습에 대해 이야기하기 위해 사용됩니다.이 방법은 깊이의 새로운 차원을 유도하는 중첩이라는 개념을 도입합니다.

Instead of adding a layer after another, the idea is to add several layers inside another one.
다른 레이어를 추가하는 대신 다른 레이어 안에 여러 레이어를 추가하는 것이 좋습니다.

It’s abstract enough.
충분히 추상적입니다.

But when you take a closer look, it’s not that complicated.
그러나 당신이 더 자세히 볼 때, 그것은 그렇게 복잡하지 않습니다.

The heart of method is divided into two parts: The PrimaryCapsules and the DigitCaps.
방법의 핵심은 PrimaryCapsules와 DigitCaps의 두 부분으로 나뉩니다.

In our case, this last will be renamed the TrafficSignCaps.
우리의 경우이 마지막 이름은 TrafficSignCaps로 이름이 바뀝니다.

**Comment by sweaterr** [comment]
## Capsules and PrimaryCapsules

![](https://cdn-images-1.medium.com/max/800/1*BW5myj5vhLgnPjbLniOZZw.png)

This layer is based on a classical convolution and creates a new convolution above, made of N*C filters.
이 레이어는 고전적인 컨볼 루션을 기반으로하며 N * C 필터로 구성된 새로운 컨볼 루션을 만듭니다.

With N, the number of filters and C the size of each capsule.
N으로, 필터의 수와 C는 각 캡슐의 크기입니다.

Thus N*C new images with (T,T) size are created.
따라서 (T, T) 크기의 N * C 개의 새로운 이미지가 생성됩니다.

In the above image, the values ​​of each capsule are shown in red in the newly created images.
위 이미지에서 새로 생성 된 이미지에는 각 캡슐의 값이 빨간색으로 표시됩니다.

This is summarized by the following code:.
이것은 다음 코드로 요약됩니다.

**Comment by sweaterr** [comment]
```python
conv = tf.contrib.layers.conv2d(input_layer, N * C, kernel, stride, padding="VALID")
# Shape: (?, T, T, N * C)
```

**Comment by sweaterr** [comment]
Now that convolution is created, we can rearrange the convolution for creating our capsules:.
컨볼 루션이 생성되었으므로 캡슐 작성을위한 컨볼 루션을 재정렬 할 수 있습니다.

```python
capsules = tf.reshape(conv, shape=(-1, T*T*N, C, 1))
# Shape: (?, T*T*N, C, 1)

# conv[0][0][0][:C] <=> capsules[0]
```

**Comment by sweaterr** [comment]
We then find ourselves with T*T*N (1152 capsules in this project case) capsules with C size.
그런 다음 T * T * N (이 프로젝트의 경우 1152 캡슐) C 크기의 캡슐을 사용합니다.

Let’s note that the first C values of convolution (see comment in the code) are equal to the values ​​of the first capsule.
convolution (코드의 주석 참조)의 첫 번째 C 값은 첫 번째 캡슐의 값과 같습니다.

Just like the above code illustrates it.
위의 코드와 마찬가지로이를 보여줍니다.

Finally, the paper offers us a new non-linear function to apply individually on each capsule.
마지막으로이 논문은 각 캡슐에 개별적으로 적용 할 수있는 새로운 비선형 함수를 제공합니다.

This new function is called “squashing”, and looks like this:.
이 새로운 기능은 "squashing"이라고 불리며 다음과 같습니다.

**Comment by sweaterr** [comment]
## TrafficSignCaps

![](https://cdn-images-1.medium.com/max/800/1*RMFmqbz9f6UWaqj80wfSHA.png)

The TrafficSignCaps layer: This layer is composed of 43 capsules in my case, each representing a certain road sign.
TrafficSignCaps 레이어 :이 레이어는 내 케이스에 43 개의 캡슐로 구성되어 있으며 각각은 특정 도로 표지를 나타냅니다.

In order to determine the model’s prediction, one chooses the capsule having the highest length.
모델의 예측을 결정하기 위해, 가장 긴 캡슐을 선택한다.

But before that, the goal will be to make the transition between the 1152 capsules of the previous layer.
그러나 그 전에 목표는 이전 층의 1152 캡슐 사이를 전환하는 것입니다.

This will be done by the method of “routing”.
이것은 "라우팅"방법으로 수행됩니다.

The purpose of that method is to choose which capsules from the previous layer must be associated with the output layer capsules.
이 방법의 목적은 이전 레이어의 캡슐을 출력 레이어 캡슐과 연결해야하는지 여부를 선택하는 것입니다.

In other words, for each capsule, a new neural network asks itself “Hey, is this capsule interesting for this class?”.
다시 말해, 각 캡슐에 대해 새로운 신경망이 "이 클래스는이 캡슐이 흥미 롭습니까?"라고 묻습니다.

**Comment by sweaterr** [comment]
## Model Architecture

**Comment by sweaterr** [comment]
```python
def _build_main_network(self, images, conv_2_dropout):
    """
        This method is used to create the two convolutions and the CapsNet on the top
        **input:
            *images: Image PLaceholder
            *conv_2_dropout: Dropout value placeholder
        **return: **
            *Caps1: Output of first Capsule layer
            *Caps2: Output of second Capsule layer
    """
    # First BLock:
    # Layer 1: Convolution.
    shape = (self.h.conv_1_size, self.h.conv_1_size, 3, self.h.conv_1_nb)
    conv1 = self._create_conv(self.tf_images, shape, relu=True, max_pooling=False, padding='VALID')
    # Layer 2: Convolution.
    shape = (self.h.conv_2_size, self.h.conv_2_size, self.h.conv_1_nb, self.h.conv_2_nb)
    conv2 = self._create_conv(conv1, shape, relu=True, max_pooling=False, padding='VALID')
    conv2 = tf.nn.dropout(conv2, keep_prob=conv_2_dropout)

    # Create the first capsules layer
    caps1 = conv_caps_layer(
        input_layer=conv2,
        capsules_size=self.h.caps_1_vec_len,
        nb_filters=self.h.caps_1_nb_filter,
        kernel=self.h.caps_1_size)
    # Create the second capsules layer used to predict the output
    caps2 = fully_connected_caps_layer(
        input_layer=caps1,
        capsules_size=self.h.caps_2_vec_len,
        nb_capsules=self.NB_LABELS,
        iterations=self.h.routing_steps)

    return caps1, caps2
```

**Comment by sweaterr** [comment]
## Training

For the training pipeline, I use Keras’ ImageDataGenerator for being able to produce Data Augmentation, while reducing the portion allocated to increase during batches.
교육 파이프 라인의 경우 데이터 확대를 생성 할 수 있도록 Keras의 ImageDataGenerator를 사용하고 일괄 처리 동안 증가하도록 할당 된 부분은 줄입니다.

Results (Accuracy):

- Training: 99%
- Validation: 98%
- Test: 97%

The result is not up to what could be optimally obtained with a “classic” convolutional network.
그 결과는 "고전적인"컨볼 루션 네트워크를 통해 최적으로 얻을 수있는 것까지되지 않습니다.

However, I have spent most of my time implementing CapsNet, and less on hyperparameters tuning and images processing.
그러나, 나는 CapsNet을 구현하는 데 대부분의 시간을 보냈지 만, 하이퍼 파라미터 튜닝과 이미지 프로세싱에는 그다지 중요하지 않습니다.

To me, 97% seems to be a good result for a first shot.
나에게 97 %는 첫 번째 촬영에서 좋은 결과 인 것으로 보인다.

I am still working on it right now to improve that percentage.
나는 그 비율을 향상시키기 위해 지금 당장 노력 중이다.
